src/server/services/training.service.ts

```diff
@@ -121,14 +128,39 @@ export const createTrainingRequest = async ({
     }
   }
 
+  const eta = calcEta(
+    trainingParams.networkDim,
+    trainingParams.networkAlpha,
+    trainingParams.targetSteps,
+    baseModel
+  );
+  const price = eta !== undefined ? calcBuzzFromEta(eta) : eta;
+  if (price === undefined) {
+    throw throwBadRequestError(
+      'Could not compute Buzz price for training - please check your parameters.'
+    );
+  }
+  const account = await getUserBuzzAccount({ accountId: userId });
+  if (account.balance < price) {
+    throw throwInsufficientFundsError(
+      `You don't have enough Buzz to perform this action (required: ${price})`
+    );
+  }
+  await createBuzzTransaction({
+    fromAccountId: userId,
+    toAccountId: 0,
+    amount: price,
+    type: TransactionType.Training,
+  });
```

Comment:
What if we charge the buzz once we confirm the request to create the generation job is successful? So we won't even need to make refunds if the request fails for some reason. We can move this down below
```suggestion
```

Reply:
this is what we were discussing in the call yesterday. charging buzz after the job opens us up to race conditions / timing attacks in addition to buzz charging errors, which would all take place after the job has already been submitted.

Reply:
In that case we could have webhooks that would notify the app of status changes to the job, similar to how stripe handles different events for payments, invoice, customer, checkout, etc. I know that'd be a bigger effort but something to consider

Reply:
we'd have to discuss this further, im not sure that will solve the issue